src/Core/Optimise/Inline.hs

```diff
@@ -42,6 +42,8 @@ betaReduce = pass go where
       pure . CotLet [(var', tp, ex)] .  substitute (Map.singleton var ref) $ body
     CotTyApp (CotLam Big (var, _) body) tp ->
       pure $ substituteInTys (Map.singleton var tp) body
+    CotTyApp body@CotLam{} _ -> pure body -- TODO: This really shouldn't happen but it does
```

Comment:
Doesn't this just strip all `CotTyApp`s? Not that I'm complaining, just trying to understand the rationale.

Reply:
Uh, yes, it was generating broken code and removing inlining opportunities, and I didn't like that. It's my fault (in the type checker somewhere, probably). Feel free to remove it.

Reply:
Actually, I'm not sure we even need this anymore.

Reply:
We _do_ - and I'll be glad to remove it along with all the other types from Core. 
They're far more trouble than they're worth.

Without,
```
let const x _ = 0 ;;
let rankn (x : forall 'a. 'a -> int) = (x 1, x ());;
let main = rankn (const 0)
```
becomes
```
let { 
  main : { 1 : int, 2 : int } = 
    { {} | 1 : int = (λ(ab : w). 0 @int) (1), 2 : int = (λ(ab : w). 0 @unit) (unit) }
 }
```

which is just abhorrent.